tests/projects/test_projects.py

```diff
@@ -59,11 +59,12 @@ def test_fetch_project(local_git_repo, local_git_repo_uri):
     # Test that we correctly determine the dest directory to use when fetching a project.
     for force_tempdir, uri in [(True, TEST_PROJECT_DIR), (False, GIT_PROJECT_URI)]:
         dest_dir = mlflow.projects._fetch_project(uri=uri, force_tempdir=force_tempdir)
-        assert dest_dir != uri
+        assert os.path.commonprefix([dest_dir, tempfile.gettempdir()]) == tempfile.gettempdir()
         assert os.path.exists(dest_dir)
     for force_tempdir, uri in [(None, TEST_PROJECT_DIR), (False, TEST_PROJECT_DIR)]:
         assert mlflow.projects._fetch_project(uri=uri, force_tempdir=force_tempdir) == \
                os.path.abspath(TEST_PROJECT_DIR)
+    assert mlflow.projects._should_use_temp_dir("git@localhost:%s.git" % local_git_repo)
```

Comment:
Note: I didn't want to add a test that invokes `mlflow run git@github.com:databricks/mlflow-example.git` since doing so would add additional dependencies on network / slow the build, but open to doing so. Such a test would also have not caught the bug addressed by this PR (i.e. we'd successfully fetch the project into a local directory named "git@github.com:databricks/mlflow-example.git" the first time we invoked `mlflow run`, and then fail on successive invocations).

Also: it's almost-possible to test fetching an SSH git URI by calling `_fetch_project` with a URI like "sid@localhost:/path/to/local/repo", but doing so requires a password so we perform a less direct test here.

Reply:
We should have a place for integration tests like that, though, as it is the only way to verify the end-to-end functionality. We do have some tests marked "large" today which make network calls, which is our current solution.

The test would just have to call `mlflow run` twice to satisfy your concern. We should also test for both the `git@` and `https://` forms of the Git URLs. I think we should just ahead and add the integration test rather than add more hacky "test-random-internal-methods" sorts of tests.